src/mono/mono/mini/simd-intrinsics.c

```diff
@@ -698,6 +703,33 @@ emit_sri_vector (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *fsi
 		int op = id == SN_GetLower ? OP_XLOWER : OP_XUPPER;
 		return emit_simd_ins_for_sig (cfg, klass, op, 0, arg0_type, fsig, args);
 	}
+	case SN_Equals:
+	case SN_GreaterThan:
+	case SN_GreaterThanOrEqual:
+	case SN_LessThan:
+	case SN_LessThanOrEqual: {
+		gboolean is_unsigned = type_is_unsigned (fsig->params [0]);
+		MonoInst *ins = emit_xcompare (cfg, klass, arg0_type, args [0], args [1]);
+		switch (id) {
+		case SN_Equals:
```

Comment:
Which "Equals" is this?

That is, we have `operator ==` and `Equals` (both the instance/static) and they differ in behavior.

`operator ==` implements the typical hardware equality and so for floating-point `NaN != NaN`. While `Equals` has to implement the contract laid out by `object.Equals` and `object.GetHashCode` and so `NaN.Equals(NaN)`

Reply:
This is the [per-element equality](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector64.cs#L1521) which returns another. The `==` equality is `SN_op_Equality` and it's supposed to return a single boolean. I'm working on `==` operator impl as well (and also the related `EqualsAll` and `EqualsAny`) and I'll push the commit later today.

Reply:
Question: why isn't the implementation of these methods in c#, like if (Sse2.IsSupported) <do sse2> etc.

Reply:
> Question: why isn't the implementation of these methods in c#, like if (Sse2.IsSupported) etc.

Most of these methods are considered "perf critical" and that depends on the inliner working and working "perfectly". The C# implementation of these are quite complex for something that is ideally optimized down to a single instruction and so it is much better/simpler to handle these in import and directly treat them as the relevant intrinsic.

Reply:
This also ultimately improves JIT throughput and allows a number of other optimizations or tweaks that aren't possible or feasible for C#